core/packages/contracts/src/utils/MerkleProof.sol

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+library MerkleProof {
+    /**
+     * @notice Verify that a specific leaf element is part of the Merkle Tree at a specific position in the tree
+     *
+     * The tree would have been constructed using
+     * https://paritytech.github.io/substrate/master/binary_merkle_tree/fn.merkle_root.html
+     *
+     * This implementation adapted from
+     * https://paritytech.github.io/substrate/master/binary_merkle_tree/fn.verify_proof.html
+     *
+     * @param root the root of the merkle tree
+     * @param leaf the leaf which needs to be proven
+     * @param position the position of the leaf, index starting with 0
+     * @param width the width or number of leaves in the tree
+     * @param proof the array of proofs to help verify the leaf's membership, ordered from leaf to root
+     * @return a boolean value representing the success or failure of the operation
+     */
+    function verify(bytes32 root, bytes32 leaf, uint256 position, uint256 width, bytes32[] memory proof)
+        internal
+        pure
+        returns (bool)
+    {
+        if (position >= width) {
+            return false;
+        }
+        return root == computeRoot(leaf, position, width, proof);
+    }
+
+    function computeRoot(bytes32 leaf, uint256 position, uint256 width, bytes32[] memory proof)
+        internal
+        pure
+        returns (bytes32)
+    {
```

Comment:
Nice! I was just making this change as a follow-up and got merge conflicts. Should we keep some `require`s here? Might want a `MAX_INT` constant instead of hard-coding the max value.
```suggestion
    {
        require(position < width, "Merkle position is too high");
        require(position < 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff);
```

Reply:
Also I moved said check to `verify`

Reply:
The intention was to check that the `position + 1` value in `computeRoot` doesn't overflow, but that should be implicitly covered by `position < width`